---
title: "SLiM: review and profiling for malaria models"
author: "Shazia Ruybal-PesÃ¡ntez"
date: "Last updated: `r format(Sys.Date(), '%d %b %Y')`"
output: rmarkdown::html_vignette
vignette: >
  %\VignetteIndexEntry{SLiM: review and profiling for malaria models}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE, 
                      warning = FALSE, 
                      message = FALSE, 
                      fig.width = 8, 
                      fig.height = 6)

library(here)
library(tidyverse)
library(patchwork)
library(plotly)
```

## Background

This notebook documents the review of the forward genetic simulation tool [SLiM](https://messerlab.org/slim/) [@haller2019; @haller2016; @messer2013; @haller2019]. We are testing whether this tool can be used as a part of the [SIMPLEGEN](mrc-ide.github.io/SIMPLEGEN) pipeline. Some key advantages of SLiM that would be beneficial for our purposes is the ability to simulate haploid genomes and because of its well-integrated `tskit` tree structure format for genetic data, which is already debugged and working seamlessly with simulation output from SLiM. However we will test whether SLiM is suitable for large population sizes and to build an appropriate malaria model (eg Ross-MacDonald). *(Potentially other things to test as well)*

#### **Main aims:**

-   Literature review of SLiM for malaria simulations (have people used this tool? if not, maybe there is a reason why...)
-   Attempt to build a simple *P. falciparum* genetic model
-   Attempt to build a simple Ross-MacDonald host-vector model
-   Profile SLiM for simulations with large population sizes and large generation/run times

## Brief overview of SLiM

The majority of papers citing SLiM use SLiM for simulations of human genomics because of the flexibility to simulate one or more genomic elements and different types of mutations (e.g. beneficial, deleterious, neutral) simultaneously under realistic evolutionary conditions (mutation and recombination rates). In addition, the [SLiM manual](https://messerlab.org/slim/) (\>700 pages) provides a comprehensive resource on how to customize SLiM, details on the coding language Eidos, ready-made "recipes" for common scenarios (eg selective sweeps, bottlenecks etc) and the interactive GUI enables visualization of the genome in real-time as simulations are running. For example, a particular beneficial mutation can be seen increasing in frequency until fixation. SLiM also provides useful profiling reports that document CPU and memory usage, and show the breakdown of usage based on distinct parts of the code. SLiM can be run in the interactive GUI or via the command line.

#### *Outputs from SLiM*

SLiM allows several output formats and also supports custom scripting using the Eidos language. This is possible not only for model specifications/customization but also for the model output (eg using `paste()` or `cat()` to build output data frames).

Some of the output possibilities in SLiM include (but not limited to):

-   `outputGenome()`: entire genome information

-   `outputVCF()`: nucleotide VCF format (note based on a sample of individuals and need to consider `outputMultiallelics` parameter)

-   `outputMutations()` or `outputFixedMutations()`: information on explicitly tracked or fixed mutations, respectively

-   `outputPedigree()`: pedigree tracking can be turned on and the simulation will keep track of every individual, their parents, grandparents etc. Pedigree output requires custom Eidos scripting (as far as I can tell).

-   `outputTreeSeq()`: output tree-sequence recording in `tskit`-format `.trees` (note much more efficient and more compact than eg VCF). Worth noting that `msprime` can write vcf files from `.trees`

#### *Tree-sequence recording in SLiM*

One of the key features of SLiM is the ability to output genetic data in tree-sequence format, `.trees` [@kelleher2018] to track population history and true local ancestry. This type of output then enables seamless integration with [`tskit`](https://tskit.dev/tskit/docs/stable/introduction.html) and [`msprime`](https://tskit.dev/msprime/docs/stable/intro.html) and can be analyzed downstream with Python and [`pyslim`](https://tskit.dev/pyslim/docs/stable/introduction.html), a Python API developed specifically for interchangeability between SLiM, `tskit`, `msprime` etc. Similarly R can be used downstream and there are also specific R packages such as [{slendr}](https://www.slendr.net) developed specifically for this (see [@petr2022]). There are several advantages to tree-sequence recording such as but not limited to (for more details see [@haller2019a] and SLiM manual Chapter 1.7 and 17):

-   **Overlaying neutral mutations:** forward simulations can be run without neutral mutations to significantly increase computational efficiency and speed (by order of magnitude or more if a model contains many neutral mutations). The neutral mutations can then be overlaid only on "pruned" trees using `msprime`.

-   **Increase efficiency of burn-in:** forward simulations can be run with no neutral burn-in (ie empty genomes) and then using "recapitation" the ancestry is reconstructed using the coalescent of only the ancestry trees present at the end. Neutral mutations can then be overlaid after the fact.

    -   Note: it is also possible to seamlessly move between coalescent and forward simulations, eg burn-in can be simulated in `msprime` without mutations, then the `.trees` file can be used in SLiM as the starting state (ie non-neutral simulations can be simulated forwards in time). Mutations can be overlaid either with `msprime` or after the forwards simulation as described above.

-   **Direct analysis of ancestry trees:** when the ancestry pattern is of interest (not the pattern of neutral mutations), inferences may be more robust for this purpose by using the recorded tree sequence (ie every possible mutational history event given true history of inheritance/recombination)

#### *Multi-species models*

The most recent release of SLiM, 4 [@haller2022] allows multi-species simulations and the ability for custom scripting to deal with spatial landscapes (continuous-space models). This version now also allows simulation of "non-genetic" species, enabling modeling of ecological scenarios (eg fox and mouse), with the flexibility to simulate genetics for one or more species or none, but also allowing interactions between them as needed. We explore this possibility for building a vector-host model (see [Profiling SLiM: basic Ross-Macdonald epidemiological model of malaria] section below).

#### *Note to self: some notable and useful resources/papers*

Below is a list of some useful papers citing SLiM as well as other useful resources.

-   Champer et al [@champer2022] use SLiM for simulations of *Anopheles* mosquito gene drive. Their SLIM models are available on GitHub (useful to look at): [https://github.com/jchamper/ChamperLab/blob/main/Mosquito-Drive-Modeling/](https://github.com/jchamper/ChamperLab/blob/main/Mosquito-Drive-Modeling/models/anoph_pan.slim)

-   [@matthey-doret2021] describes R wrapper {`SimBit`} that implements tree structures for output (might be useful to see how tskit is implemented in another C++/R framework)

-   [@cury2022] useful guide on how to build appropriate bacterial models in SLiM (and useful with a good overview of implemented models in SLiM with rationale for model params and coding choices etc)

    -   they use nonWF SLiM model with tree-sequence recording + recapitation in `msprime`, first run forward simulation and then use recapitation on ancestral branches to produce coalescence

-   [@cao2021] useful guide on how to build haploid simulations in SLiM

-   [@hardy2022] simulates subpops (ie demes) sing SLiM to model individual genomes, which can reproduce, mutate, recombine and die

-   [@sabin2022] use SLiM to model Mycobacterium based on nucleotide sequences, SLiM model params available on github, eg: <https://github.com/sjsabin/mcan_popgen/blob/main/simulations/base/base.slim>

-   `{stdpopsim}` ([@adrion2020]) catalog of different species and demographic history (include *Anopheles* as per Miles 2017 params)

-   `{slimr}` R package [@dinnage2021] that interfaces with SLiM 3.0: <https://github.com/rdinnager/slimr/>

    -   Downside: seems like its not compatible with SLiM 4.0

    -   Workshop recording: <https://zoom.us/rec/share/T5thlw67U8T3-BNQUJoZiaKTCSk2xeQTkyDpvOwvtUnPRO7VCE8tvJzUJsGTYnkd.xHFECYM5vDtJdp12> (passcode: uk5\^G\$z7)

    -   Workshop code: <https://github.com/rdinnager/slimr_workshop_CBA>

-   `fwddpp` (C++ template library) [@thornton2014] orders of magnitude faster than SLiM - haploids??

    -   use cases: large Ns and selection

    -   fastest algorithm

    -   nonstandard fitness models and/or modeling fitness-to-fitness relationships

    -   maximize runtime efficiency for particular demo scenario

-   Some useful examples of how others have used SLiM and documented their workflows are below:

    -   [@blischak2020] exported individual genotype data from each simulation as vcfs (`.outputVCFSample()`, they also have publicly available code for their SLiM models implemented in python (<https://github.com/pblischak/inbreeding-sfs/tree/master/sims/SLiM>)

    -   [@anderson2021] provides useful a description of SLiM as part of larger pipeline (the SLiM output can be used downstream using R package {`CKMRpop}`). *Note: the output described here seems very similar to the transmission record of IDs produced by SIMPLEGEN.*
