<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>SLiM: review and profiling for malaria models • SIMPLEGEN</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link href="https://cdnjs.cloudflare.com/ajax/libs/bootswatch/3.4.0/flatly/bootstrap.min.css" rel="stylesheet" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><link href="../extra.css" rel="stylesheet">
<meta property="og:title" content="SLiM: review and profiling for malaria models">
<meta property="og:description" content="SIMPLEGEN">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">
    

    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">SIMPLEGEN</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">0.2.0</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Get started
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/installation.html">Installation</a>
    </li>
    <li class="divider">
    </li>
<li class="dropdown-header">Simulating malaria transmission</li>
    <li>
      <a href="../404.html">A simple model</a>
    </li>
    <li>
      <a href="../404.html">Population surveys</a>
    </li>
    <li class="divider">
    </li>
<li class="dropdown-header">Overlaying genetics</li>
    <li>
      <a href="../404.html">Pruning the transmission record</a>
    </li>
    <li>
      <a href="../404.html">Pedigrees and relatedness</a>
    </li>
  </ul>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    The models
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li class="dropdown-header">Detailed descriptions</li>
    <li>
      <a href="../articles/model_description.html">The inbuilt transmission model</a>
    </li>
    <li class="divider">
    </li>
<li class="dropdown-header">Model calibration</li>
    <li>
      <a href="../articles/calibration_approach.html">Overall approach</a>
    </li>
    <li>
      <a href="../articles/calibration_prev_inc.html">Prevalence, incidence and EIR</a>
    </li>
    <li>
      <a href="../articles/calibration_diagnostics.html">Diagnostic sensitivity</a>
    </li>
    <li>
      <a href="../articles/calibration_age_inc.html">Age-incidence</a>
    </li>
  </ul>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    The data
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li class="dropdown-header">Epidemiological data</li>
    <li>
      <a href="../articles/data_prev_inc.html">Prevalence, incidence and EIR</a>
    </li>
    <li>
      <a href="../articles/data_diagnostics.html">Diagnostic sensitivity</a>
    </li>
    <li>
      <a href="../articles/data_age_inc.html">Age-incidence</a>
    </li>
    <li class="divider">
    </li>
<li class="dropdown-header">Genetic data</li>
    <li>
      <a href="../404.html">Complexity of infection</a>
    </li>
  </ul>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Misc
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/SLiM_review.html">Integrating with SLiM</a>
    </li>
  </ul>
</li>
<li>
  <a href="../reference/index.html">Functions</a>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/mrc-ide/SIMPLEGEN" class="external-link">
    <span class="fa fa-github"></span>
     
  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>SLiM: review and profiling for malaria
models</h1>
                        <h4 data-toc-skip class="author">Shazia
Ruybal-Pesántez</h4>
            
            <h4 data-toc-skip class="date">Last updated: 28 Mar
2023</h4>
      
      <small class="dont-index">Source: <a href="https://github.com/mrc-ide/SIMPLEGEN/blob/HEAD/vignettes/SLiM_review.Rmd" class="external-link"><code>vignettes/SLiM_review.Rmd</code></a></small>
      <div class="hidden name"><code>SLiM_review.Rmd</code></div>

    </div>

    
    
<div class="section level2">
<h2 id="background">Background<a class="anchor" aria-label="anchor" href="#background"></a>
</h2>
<p>This notebook documents the review of the forward genetic simulation
tool <a href="https://messerlab.org/slim/" class="external-link">SLiM</a> <span class="citation">[@haller2019; @haller2016; @messer2013;
@haller2019]</span>. We are testing whether this tool can be used as a
part of the <a href="mrc-ide.github.io/SIMPLEGEN">SIMPLEGEN</a>
pipeline. Some key advantages of SLiM that would be beneficial for our
purposes is the ability to simulate haploid genomes and because of its
well-integrated <code>tskit</code> tree structure format for genetic
data, which is already debugged and working seamlessly with simulation
output from SLiM. However we will test whether SLiM is suitable for
large population sizes and to build an appropriate malaria model (eg
Ross-MacDonald). <em>(Potentially other things to test as well)</em></p>
<div class="section level4">
<h4 id="main-aims">
<strong>Main aims:</strong><a class="anchor" aria-label="anchor" href="#main-aims"></a>
</h4>
<ul>
<li>Literature review of SLiM for malaria simulations (have people used
this tool? if not, maybe there is a reason why…)</li>
<li>Attempt to build a simple <em>P. falciparum</em> genetic model</li>
<li>Attempt to build a simple Ross-MacDonald host-vector model</li>
<li>Profile SLiM for simulations with large population sizes and large
generation/run times</li>
</ul>
</div>
</div>
<div class="section level2">
<h2 id="brief-overview-of-slim">Brief overview of SLiM<a class="anchor" aria-label="anchor" href="#brief-overview-of-slim"></a>
</h2>
<p>The majority of papers citing SLiM use SLiM for simulations of human
genomics because of the flexibility to simulate one or more genomic
elements and different types of mutations (e.g. beneficial, deleterious,
neutral) simultaneously under realistic evolutionary conditions
(mutation and recombination rates). In addition, the <a href="https://messerlab.org/slim/" class="external-link">SLiM manual</a> (&gt;700 pages)
provides a comprehensive resource on how to customize SLiM, details on
the coding language Eidos, ready-made “recipes” for common scenarios (eg
selective sweeps, bottlenecks etc) and the interactive GUI enables
visualization of the genome in real-time as simulations are running. For
example, a particular beneficial mutation can be seen increasing in
frequency until fixation. SLiM also provides useful profiling reports
that document CPU and memory usage, and show the breakdown of usage
based on distinct parts of the code. SLiM can be run in the interactive
GUI or via the command line.</p>
<div class="section level4">
<h4 id="outputs-from-slim">
<em>Outputs from SLiM</em><a class="anchor" aria-label="anchor" href="#outputs-from-slim"></a>
</h4>
<p>SLiM allows several output formats and also supports custom scripting
using the Eidos language. This is possible not only for model
specifications/customization but also for the model output (eg using
<code><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste()</a></code> or <code><a href="https://rdrr.io/r/base/cat.html" class="external-link">cat()</a></code> to build output data
frames).</p>
<p>Some of the output possibilities in SLiM include (but not limited
to):</p>
<ul>
<li><p><code>outputGenome()</code>: entire genome information</p></li>
<li><p><code>outputVCF()</code>: nucleotide VCF format (note based on a
sample of individuals and need to consider
<code>outputMultiallelics</code> parameter)</p></li>
<li><p><code>outputMutations()</code> or
<code>outputFixedMutations()</code>: information on explicitly tracked
or fixed mutations, respectively</p></li>
<li><p><code>outputPedigree()</code>: pedigree tracking can be turned on
and the simulation will keep track of every individual, their parents,
grandparents etc. Pedigree output requires custom Eidos scripting (as
far as I can tell).</p></li>
<li><p><code>outputTreeSeq()</code>: output tree-sequence recording in
<code>tskit</code>-format <code>.trees</code> (note much more efficient
and more compact than eg VCF). Worth noting that <code>msprime</code>
can write vcf files from <code>.trees</code></p></li>
</ul>
</div>
<div class="section level4">
<h4 id="tree-sequence-recording-in-slim">
<em>Tree-sequence recording in SLiM</em><a class="anchor" aria-label="anchor" href="#tree-sequence-recording-in-slim"></a>
</h4>
<p>One of the key features of SLiM is the ability to output genetic data
in tree-sequence format, <code>.trees</code> <span class="citation">[@kelleher2018]</span> to track population history and
true local ancestry. This type of output then enables seamless
integration with <a href="https://tskit.dev/tskit/docs/stable/introduction.html" class="external-link"><code>tskit</code></a>
and <a href="https://tskit.dev/msprime/docs/stable/intro.html" class="external-link"><code>msprime</code></a>
and can be analyzed downstream with Python and <a href="https://tskit.dev/pyslim/docs/stable/introduction.html" class="external-link"><code>pyslim</code></a>,
a Python API developed specifically for interchangeability between SLiM,
<code>tskit</code>, <code>msprime</code> etc. Similarly R can be used
downstream and there are also specific R packages such as <a href="https://www.slendr.net" class="external-link">{slendr}</a> developed specifically for
this (see <span class="citation">[@petr2022]</span>). There are several
advantages to tree-sequence recording such as but not limited to (for
more details see <span class="citation">[@haller2019a]</span> and SLiM
manual Chapter 1.7 and 17):</p>
<ul>
<li><p><strong>Overlaying neutral mutations:</strong> forward
simulations can be run without neutral mutations to significantly
increase computational efficiency and speed (by order of magnitude or
more if a model contains many neutral mutations). The neutral mutations
can then be overlaid only on “pruned” trees using
<code>msprime</code>.</p></li>
<li>
<p><strong>Increase efficiency of burn-in:</strong> forward
simulations can be run with no neutral burn-in (ie empty genomes) and
then using “recapitation” the ancestry is reconstructed using the
coalescent of only the ancestry trees present at the end. Neutral
mutations can then be overlaid after the fact.</p>
<ul>
<li>Note: it is also possible to seamlessly move between coalescent and
forward simulations, eg burn-in can be simulated in <code>msprime</code>
without mutations, then the <code>.trees</code> file can be used in SLiM
as the starting state (ie non-neutral simulations can be simulated
forwards in time). Mutations can be overlaid either with
<code>msprime</code> or after the forwards simulation as described
above.</li>
</ul>
</li>
<li><p><strong>Direct analysis of ancestry trees:</strong> when the
ancestry pattern is of interest (not the pattern of neutral mutations),
inferences may be more robust for this purpose by using the recorded
tree sequence (ie every possible mutational history event given true
history of inheritance/recombination)</p></li>
</ul>
</div>
<div class="section level4">
<h4 id="multi-species-models">
<em>Multi-species models</em><a class="anchor" aria-label="anchor" href="#multi-species-models"></a>
</h4>
<p>The most recent release of SLiM, 4 <span class="citation">[@haller2022]</span> allows multi-species simulations
and the ability for custom scripting to deal with spatial landscapes
(continuous-space models). This version now also allows simulation of
“non-genetic” species, enabling modeling of ecological scenarios (eg fox
and mouse), with the flexibility to simulate genetics for one or more
species or none, but also allowing interactions between them as needed.
We explore this possibility for building a vector-host model (see
[Profiling SLiM: basic Ross-Macdonald epidemiological model of malaria]
section below).</p>
</div>
<div class="section level4">
<h4 id="note-to-self-some-notable-and-useful-resourcespapers">
<em>Note to self: some notable and useful resources/papers</em><a class="anchor" aria-label="anchor" href="#note-to-self-some-notable-and-useful-resourcespapers"></a>
</h4>
<p>Below is a list of some useful papers citing SLiM as well as other
useful resources.</p>
<ul>
<li><p>Champer et al <span class="citation">[@champer2022]</span> use
SLiM for simulations of <em>Anopheles</em> mosquito gene drive. Their
SLIM models are available on GitHub (useful to look at): <a href="https://github.com/jchamper/ChamperLab/blob/main/Mosquito-Drive-Modeling/models/anoph_pan.slim" class="external-link">https://github.com/jchamper/ChamperLab/blob/main/Mosquito-Drive-Modeling/</a></p></li>
<li><p><span class="citation">[@matthey-doret2021]</span> describes R
wrapper {<code>SimBit</code>} that implements tree structures for output
(might be useful to see how tskit is implemented in another C++/R
framework)</p></li>
<li>
<p><span class="citation">[@cury2022]</span> useful guide on how to
build appropriate bacterial models in SLiM (and useful with a good
overview of implemented models in SLiM with rationale for model params
and coding choices etc)</p>
<ul>
<li>they use nonWF SLiM model with tree-sequence recording +
recapitation in <code>msprime</code>, first run forward simulation and
then use recapitation on ancestral branches to produce coalescence</li>
</ul>
</li>
<li><p><span class="citation">[@cao2021]</span> useful guide on how to
build haploid simulations in SLiM</p></li>
<li><p><span class="citation">[@hardy2022]</span> simulates subpops (ie
demes) sing SLiM to model individual genomes, which can reproduce,
mutate, recombine and die</p></li>
<li><p><span class="citation">[@sabin2022]</span> use SLiM to model
Mycobacterium based on nucleotide sequences, SLiM model params available
on github, eg: <a href="https://github.com/sjsabin/mcan_popgen/blob/main/simulations/base/base.slim" class="external-link uri">https://github.com/sjsabin/mcan_popgen/blob/main/simulations/base/base.slim</a></p></li>
<li><p><code>{stdpopsim}</code> (<span class="citation">[@adrion2020]</span>) catalog of different species and
demographic history (include <em>Anopheles</em> as per Miles 2017
params)</p></li>
<li>
<p><code>{slimr}</code> R package <span class="citation">[@dinnage2021]</span> that interfaces with SLiM 3.0: <a href="https://github.com/rdinnager/slimr/" class="external-link uri">https://github.com/rdinnager/slimr/</a></p>
<ul>
<li><p>Downside: seems like its not compatible with SLiM 4.0</p></li>
<li><p>Workshop recording: <a href="https://zoom.us/rec/share/T5thlw67U8T3-BNQUJoZiaKTCSk2xeQTkyDpvOwvtUnPRO7VCE8tvJzUJsGTYnkd.xHFECYM5vDtJdp12" class="external-link uri">https://zoom.us/rec/share/T5thlw67U8T3-BNQUJoZiaKTCSk2xeQTkyDpvOwvtUnPRO7VCE8tvJzUJsGTYnkd.xHFECYM5vDtJdp12</a>
(passcode: uk5^G$z7)</p></li>
<li><p>Workshop code: <a href="https://github.com/rdinnager/slimr_workshop_CBA" class="external-link uri">https://github.com/rdinnager/slimr_workshop_CBA</a></p></li>
</ul>
</li>
<li>
<p><code>fwddpp</code> (C++ template library) <span class="citation">[@thornton2014]</span> orders of magnitude faster than
SLiM - haploids??</p>
<ul>
<li><p>use cases: large Ns and selection</p></li>
<li><p>fastest algorithm</p></li>
<li><p>nonstandard fitness models and/or modeling fitness-to-fitness
relationships</p></li>
<li><p>maximize runtime efficiency for particular demo scenario</p></li>
</ul>
</li>
<li>
<p>Some useful examples of how others have used SLiM and documented
their workflows are below:</p>
<ul>
<li><p><span class="citation">[@blischak2020]</span> exported individual
genotype data from each simulation as vcfs
(<code>.outputVCFSample()</code>, they also have publicly available code
for their SLiM models implemented in python (<a href="https://github.com/pblischak/inbreeding-sfs/tree/master/sims/SLiM" class="external-link uri">https://github.com/pblischak/inbreeding-sfs/tree/master/sims/SLiM</a>)</p></li>
<li><p><span class="citation">[@anderson2021]</span> provides useful a
description of SLiM as part of larger pipeline (the SLiM output can be
used downstream using R package {<code>CKMRpop}</code>). <em>Note: the
output described here seems very similar to the transmission record of
IDs produced by SIMPLEGEN.</em></p></li>
</ul>
</li>
</ul>
</div>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

        <nav id="toc" data-toggle="toc"><h2 data-toc-skip>Contents</h2>
    </nav>
</div>

</div>



      <footer><div class="copyright">
  <p></p>
<p>Developed by Bob Verity, Shazia Ruybal, Isobel Routledge, Sophie Berube, Daniel Larremore.</p>
</div>

<div class="pkgdown">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.7.</p>
</div>

      </footer>
</div>

  


  

  </body>
</html>
