<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>What is the SIMPLEGEN pipeline? • SIMPLEGEN</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link href="https://cdnjs.cloudflare.com/ajax/libs/bootswatch/3.4.0/flatly/bootstrap.min.css" rel="stylesheet" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><link href="../extra.css" rel="stylesheet">
<meta property="og:title" content="What is the SIMPLEGEN pipeline?">
<meta property="og:description" content="SIMPLEGEN">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">
    

    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">SIMPLEGEN</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">0.2.0</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Get started
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/installation.html">Installation</a>
    </li>
    <li class="divider">
    </li>
<li class="dropdown-header">Simulating malaria transmission</li>
    <li>
      <a href="../404.html">TODO - A simple model</a>
    </li>
    <li>
      <a href="../404.html">TODO - Sampling from the population</a>
    </li>
    <li class="divider">
    </li>
<li class="dropdown-header">Overlaying genetics</li>
    <li>
      <a href="../404.html">TODO - Pruning the transmission record</a>
    </li>
    <li>
      <a href="../404.html">TODO - Pedigrees and relatedness</a>
    </li>
  </ul>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    The pipeline
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/what_is_simplegen.html">What is the SIMPLEGEN pipeline?</a>
    </li>
    <li>
      <a href="../articles/transmission_record.html">The transmission record</a>
    </li>
  </ul>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Models and calibration
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li class="dropdown-header">Overall approach</li>
    <li>
      <a href="../articles/calibration_approach.html">Model development and calibration</a>
    </li>
    <li class="divider">
    </li>
<li class="dropdown-header">Transmission models</li>
    <li>
      <a href="../articles/model_description.html">The inbuilt transmission model</a>
    </li>
    <li class="dropdown-submenu">
      <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">Calibration</a>
      <ul class="dropdown-menu" role="menu">
<li>
          <a href="../articles/calibration_intro.html">Calibrating the in-built transmission model</a>
        </li>
        <li>
          <a href="../articles/calibration_prev_inc.html">Prevalence, incidence and EIR</a>
        </li>
        <li>
          <a href="../articles/calibration_diagnostics.html">Diagnostics</a>
        </li>
        <li>
          <a href="../articles/calibration_age_inc.html">Age-incidence</a>
        </li>
      </ul>
</li>
    <li class="divider">
    </li>
<li class="dropdown-header">Genetic models</li>
    <li>
      <a href="../404.html">TODO - Description</a>
    </li>
    <li class="dropdown-submenu">
      <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">Calibration</a>
      <ul class="dropdown-menu" role="menu">
<li>
          <a href="../404.html">TODO - Genetic calibration</a>
        </li>
      </ul>
</li>
  </ul>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Data
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li class="dropdown-header">Epidemiological data</li>
    <li>
      <a href="../articles/data_prev_inc.html">Prevalence, incidence and EIR</a>
    </li>
    <li>
      <a href="../articles/data_diagnostics.html">Diagnostic sensitivity</a>
    </li>
    <li>
      <a href="../articles/data_age_inc.html">Age-incidence</a>
    </li>
    <li class="divider">
    </li>
<li class="dropdown-header">Genetic data</li>
    <li>
      <a href="../404.html">TODO - Complexity of infection</a>
    </li>
  </ul>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Misc
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/SLiM_review.html">Integrating with SLiM</a>
    </li>
  </ul>
</li>
<li>
  <a href="../reference/index.html">Functions</a>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/mrc-ide/SIMPLEGEN" class="external-link">
    <span class="fa fa-github"></span>
     
  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>What is the SIMPLEGEN pipeline?</h1>
                        <h4 data-toc-skip class="author">Bob Verity</h4>
            
            <h4 data-toc-skip class="date">Last updated: 30 Mar
2023</h4>
      
      <small class="dont-index">Source: <a href="https://github.com/mrc-ide/SIMPLEGEN/blob/HEAD/vignettes/what_is_simplegen.Rmd" class="external-link"><code>vignettes/what_is_simplegen.Rmd</code></a></small>
      <div class="hidden name"><code>what_is_simplegen.Rmd</code></div>

    </div>

    
    
<div class="section level3">
<h3 id="what-is-malaria-molecular-surveillance">What is malaria molecular surveillance?<a class="anchor" aria-label="anchor" href="#what-is-malaria-molecular-surveillance"></a>
</h3>
<p>Malaria molecular surveillance (MMS), sometimes called “genomic
surveillance”, is a powerful approach for understanding what is going on
inside a malaria-infected population. When we sequence the genome of
<em>Plasmodium</em> parasites inside the blood of infected individuals,
the information we get back can tell us something about the history of
infections. For example, if we see two people carrying parasites with
almost identical genomes then it is likely there is a short chain of
infections connecting them. This is because genotypes undergo
recombination inside the mosquito, meaning they tend to get shuffled
together over subsequent infection events. Similarly, when we sequence a
single person we might find evidence of not just one parasite genotype,
but a whole population of multiple different genotypes (a high
“complexity of infection”, COI). This is a clue that the person may have
been bitten multiple times by different mosquitoes, and so this tells us
something about transmission intensity in the region.</p>
<p>However, both of these examples are more complicated than they first
seem. In low-transmission areas we would expect to see many more pairs
of people with similar parasite genotypes than we would at high
transmission, and so we need to be careful when we start inferring
transmission chains. Likewise in the case of the COI analysis, if
somebody is infected with multiple parasite genotypes then this doesn’t
have to be because they recieved multiple mosquito bites, rather a
single mosquito may have passed on several genotypes simultaneously.</p>
<p>In MMS we try to balance between different possible explanations for
the observed genetic data so that we can extract useful information
while also ensuring that our conclusions are robust. For this reason,
MMS tends to be a highly quantitative discipline with a heavy reliance
on statistical models. This makes the analysis plan less straightforward
then simple prevalence or incidence measures, but at the same time we
obtain a fundamentally different type of information that can be very
useful for some malaria control questions.</p>
</div>
<div class="section level3">
<h3 id="why-do-we-need-to-simulate-plasmodium-genetics">Why do we need to simulate <em>Plasmodium</em> genetics?<a class="anchor" aria-label="anchor" href="#why-do-we-need-to-simulate-plasmodium-genetics"></a>
</h3>
<p>Let’s say I’ve convinced you that MMS is a potentially useful avenue
to persue, and you want to start sequencing parasites from your study
region. The problem becomes; how should I design my MMS study? There are
a huge number of variables to pin down, including the number of
clusters, the number of samples per cluster, the method of blood
extraction, the method of amplification and sequencing, the number and
distribution of loci, all the way down to the final analysis tools that
will be used to generate results. Long analysis pipelines like this have
multiple issues that can jeapardise a project. First, with long
pipelines come more points of failure. Although we might be confident
that our samples <em>can</em> flow from one step to the next, in reality
we should be prepared for some loss of information at every step (for
example samples failing QC), meaning we are left with a reduced dataset
at the end. Second, it can be difficult to predict how design changes
early on in the pipeline will influence results later on. It may be that
changing the number of targetted loci has a major effect on which
software tools can ultimately be used. And finally, bringing this
together with some statistical principles, it can be very difficult to
<em>power</em> a study - by which we mean, design a study in such a way
that we have a good chance of detecting something interesting if it is
there. For simple prevalence and incidence studies there are formulae
and tables in text books that can be used to power a study, but for
complex MMS pipelines no such simple formulae exist. In the worst case
scenario this may mean we are almost guaranteed to get a non-significant
result even before starting the study, which is a rather heartbreaking
situation to be in, especially after going to all the effort of
conducting field and lab work!</p>
<p>As a concrete example, let’s imagine you want to explore how
fragmented the parasite population is in your region using the software
<a href="https://web.stanford.edu/group/pritchardlab/structure.html" class="external-link">STRUCTURE</a>.
This software requires that you know the COI of your samples, meaning
you probably want to run something like <a href="https://github.com/EPPIcenter/THEREALMcCOIL" class="external-link">THE REAL McCOIL</a>
beforehand to estimate COI (notice how we are chaining analysis methods
together here, which is common in this field). If the number or
diversity of loci is too low then you may struggle to reliably estimate
COI greater than about 2 using THE REAL McCOIL. Therefore, in high
transmission sites where polyclonal infections are common, you can
expect to throw away upwards of 50% of your data before carrying out the
STRUCTURE analysis. If you had anticipated this beforehand then you
could mitigate by either 1) increasing the number of samples so enough
remain even after discarding, or 2) increased the number of loci so COI
can be inferred more accurately. But if not then you may be left with
very weak evidence of population structure that doesn’t really answer
the question and instead begs for a further study with a larger sample
size.</p>
<p>Simulation offers a powerful route out of this problem. By starting
with “synthetic data”, meaning data of exactly the same dimension and
file format as the real data but produced from a computer simulation
model, we have the opportunity for a trial run at our entire analysis
pipeline. Some of the simple points of failure will be identified
immediately by this method, for example incompatible file formats, but
we can also push this idea further. For example, by simulating from an
epidemiological transmission model that does a reasonable job of
capturing the sorts of dynamics we expect in our region, we can
hopefully produce synthetic data that has a similar composition in terms
of e.g. number of polyclonal infections or relatedness to our final
data. We can also make our results more realistic by using an
“observation model”, which includes things like introducing sequencing
error, random sequencing failures (i.e. missing data) and merging mixed
infections. This gives us a better idea of the sort of signal we can
expect to see in the real data if we analyse it exactly the same way.
Finally, if we want to be completely thorough, we can simulate data
under a model with the known interesting effect of interest
(i.e. population structure) and then we can systematically explore how
likely we are to detect this effect with different study design choices
- for example obtaining more samples vs. more loci. This gives us an
estimate of “empirical power” that can be used to justify a particular
study design.</p>
</div>
<div class="section level3">
<h3 id="cant-i-just-simulate-parasite-genotypes-at-the-same-time-as-simulating-transmission-events">Can’t I just simulate parasite genotypes at the same time as
simulating transmission events?<a class="anchor" aria-label="anchor" href="#cant-i-just-simulate-parasite-genotypes-at-the-same-time-as-simulating-transmission-events"></a>
</h3>
<p>Yes, you can! But for many applications this turns out to be quite
inefficient. Some of the most interesting MMS questions actually revolve
around <em>neutral</em> genetic data, by which we mean anything that is
not under selection. This rules out things like loci responsible for
drug resistance, but it includes other large swathes of the genome. If a
locus is selectively neutral then <em>by definition</em> it does not
interact with the transmission process, in other words, a parasite
strain will have the same chance of being transmitted irrespective of
the allele at this locus. From a simulation point of view, this means we
can completely decouple the process of epidemiological simulation from
the process of genetic simulation, i.e. we can simulate transmission
first and then overlay genetics afterwards. Compare this with the naive
approach of simulating genotypes at the same time as transmission - we
would have to pass genotypes back and forth between mosquitoes and human
hosts countless times, not to mention simulating recombination in the
mosquito midgut, and still the vast majority of these mosquitoes would
die before passing on infection. For this reason, we expect decoupled
simulation to be faster than joint simulation most of the time, and it
achieves its greatest advantage when the final sample is small relative
to the total population.</p>
<p>This is not to say that forwards simulation of genotypes is always a
bad idea, and indeed for loci under selection it may be the simplest
approach. There are several, excellent simulators of this sort that have
been published in recent years. But if our aim is to analyse neutral
genetic markers in large populations then SIMPLEGEN is likely to be
competitive in terms of speed.</p>
</div>
<div class="section level3">
<h3 id="how-does-simplegen-solve-this-problem">How does SIMPLEGEN solve this problem?<a class="anchor" aria-label="anchor" href="#how-does-simplegen-solve-this-problem"></a>
</h3>
<p>SIMPLEGEN is a simulation <em>pipeline</em>. It breaks down the
process described above into a series of distinct steps with
well-defined inputs and outputs. Briefly, these steps are:</p>
<ul>
<li>Simulate from an epidemiological model and output a “transmission
record” that stores key events. Also sample from the population using a
particular survey design.</li>
<li>Using the simulated sample from the population, prune the
transmission record to keep only the events that are needed.</li>
<li>Using a genetic model, simulate a pedigree describing the
relatedness between all parasite strains within all sampled
individuals.</li>
<li>Specify the physical locations along the genome that you are
interested in, and simulate blocks of relatedness using a recombination
model.</li>
<li>Specify a mutation model and simulate mutations.</li>
<li>Use an observation model to convert “true” genotypes into “observed”
genotypes that capture some of the issues in real data</li>
<li>Pass simulated data into downstream tools for analysis.</li>
</ul>
<p>The pipeline can be accessed at any point, for example, if you
already have a model that produces results in the pedigree (.ped) format
in step 4 then you can jump in at this stage. However, it is anticipated
that most people will run through from start to finish. A series of
models are distributed with the pipeline, but in general it is intended
that anyone can use any model as long as it produces results in
compatible formats.</p>
</div>
<div class="section level3">
<h3 id="what-will-i-be-able-to-do-with-simplegen">What will I be able to <em>do</em> with SIMPLEGEN?<a class="anchor" aria-label="anchor" href="#what-will-i-be-able-to-do-with-simplegen"></a>
</h3>
<p>This may still feel a little abstract, so here are some concrete
examples of things you will be able to do with the SIMPLEGEN
pipeline:</p>
<ul>
<li>simulate genetic datasets in a range of file formats, and follow
tutorials on how to analyse these datasets with downstream tools</li>
<li>work out which tools you want to chain together on your real data
analysis, i.e. come up with a detailed analysis plan</li>
<li>get a sense of the final end results of your analysis plan, for
example plots and visualisations</li>
<li>explore how much <em>power</em> you will have to detect a signal of
interest. Try different sizes of data and choices of software tools to
maximise power</li>
<li>browse publicly available datasets showing key epidemiological and
genetic relationships</li>
<li>calibrate my own models against data</li>
</ul>
</div>
<div class="section level3">
<h3 id="how-do-i-actually-use-the-pipeline">How do I actually use the pipeline?<a class="anchor" aria-label="anchor" href="#how-do-i-actually-use-the-pipeline"></a>
</h3>
<p>SIMPLEGEN is an R package that uses C++ under the hood. Start by <a href="https://mrc-ide.github.io/SIMPLEGEN/articles/installation.html" class="external-link">installing
the package</a>, then follow the series of <a href="https://mrc-ide.github.io/SIMPLEGEN/404.html" class="external-link">tutorials</a> to get
up and running.</p>
</div>
<div class="section level3">
<h3 id="arent-there-other-tools-that-do-this-already">Aren’t there other tools that do this already?<a class="anchor" aria-label="anchor" href="#arent-there-other-tools-that-do-this-already"></a>
</h3>
<p>Yes and no. The ideas underpinning SIMPLEGEN are definitely not new,
in fact the idea of efficient tree recording and decoupling simulation
processes is a growing area in population genetics with some fantastic
recent developments. However, in our <a href="https://mrc-ide.github.io/SIMPLEGEN/articles/SLiM_review.html" class="external-link">search
of similar tools</a> we have not found anything out there that is
ideally suited for use in malaria. The existing tools have a slightly
different focus, to the extent that even if they <em>could</em> be used
to do something similar to what we are proposing, the activation energy
for the user would be much higher.</p>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

      </div>

</div>



      <footer><div class="copyright">
  <p></p>
<p>Developed by Bob Verity, Shazia Ruybal, Isobel Routledge, Sophie Berube, Daniel Larremore.</p>
</div>

<div class="pkgdown">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.7.</p>
</div>

      </footer>
</div>

  


  

  </body>
</html>
