<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1.0"><title>24-SNP barcode data at two time-points in Senegal (Bei et al., 2018) — Bei_2018 • SIMPLEGEN</title><!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link href="https://cdnjs.cloudflare.com/ajax/libs/bootswatch/3.4.0/flatly/bootstrap.min.css" rel="stylesheet" crossorigin="anonymous"><script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css"><script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous"><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous"><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet"><script src="../pkgdown.js"></script><link href="../extra.css" rel="stylesheet"><meta property="og:title" content="24-SNP barcode data at two time-points in Senegal (Bei et al., 2018) — Bei_2018"><meta property="og:description" content="Data from Bei et al. (2018). Here we give a brief summary of the data - see
the original paper for full details.


Samples were randomly selected from longitudinal cohorts collected from
Dielmo/Ndiop, Senegal, and focus on 2 distant time-points (2001-2002, and
2014). The first time-point corresponds to a period of high transmission, and
the second a period of extremely low transmission. Samples were genotyped
using a 24-SNP barcode, and complexity of infection was estimated using the
COIL algorithm. Samples were also matched against a large database of
previously published and unpublished barcodes. For samples from the low
transmission period (2014), one of the three repeated barcode clusters (n =
6) corresponded to a parasite type (haplotype 3), observed in Thiès in both
2007 and 2010. The other 2 clusters (IP1, n = 2; IP2, n = 2) had not
previously been observed. Data were extracted using Tabula v1.2.1."><!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--></head><body data-spy="scroll" data-target="#toc">
    

    <div class="container template-reference-topic">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">SIMPLEGEN</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">0.2.0</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav"><li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Get started
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu"><li>
      <a href="../articles/installation.html">Installation</a>
    </li>
    <li class="divider">
    <li class="dropdown-header">Simulating malaria transmission</li>
    <li>
      <a href="../404.html">A simple model</a>
    </li>
    <li>
      <a href="../404.html">Population surveys</a>
    </li>
    <li class="divider">
    <li class="dropdown-header">Overlaying genetics</li>
    <li>
      <a href="../404.html">Pruning the transmission record</a>
    </li>
    <li>
      <a href="../404.html">Pedigrees and relatedness</a>
    </li>
  </ul></li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    The models
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu"><li class="dropdown-header">Detailed descriptions</li>
    <li>
      <a href="../articles/model_description.html">The inbuilt transmission model</a>
    </li>
    <li class="divider">
    <li class="dropdown-header">Model calibration</li>
    <li>
      <a href="../articles/calibration_approach.html">Overall approach</a>
    </li>
    <li>
      <a href="../articles/calibration_prev_inc.html">Prevalence, incidence and EIR</a>
    </li>
    <li>
      <a href="../articles/calibration_diagnostics.html">Diagnostic sensitivity</a>
    </li>
    <li>
      <a href="../articles/calibration_age_inc.html">Age-incidence</a>
    </li>
  </ul></li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    The data
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu"><li class="dropdown-header">Epidemiological data</li>
    <li>
      <a href="../articles/data_prev_inc.html">Prevalence, incidence and EIR</a>
    </li>
    <li>
      <a href="../articles/data_diagnostics.html">Diagnostic sensitivity</a>
    </li>
    <li>
      <a href="../articles/data_age_inc.html">Age-incidence</a>
    </li>
    <li class="divider">
    <li class="dropdown-header">Genetic data</li>
    <li>
      <a href="../404.html">Complexity of infection</a>
    </li>
  </ul></li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Misc
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu"><li>
      <a href="../articles/SLiM_review.html">Integrating with SLiM</a>
    </li>
  </ul></li>
<li>
  <a href="../reference/index.html">Functions</a>
</li>
      </ul><ul class="nav navbar-nav navbar-right"><li>
  <a href="https://github.com/mrc-ide/SIMPLEGEN" class="external-link">
    <span class="fa fa-github"></span>
     
  </a>
</li>
      </ul></div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      

      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>24-SNP barcode data at two time-points in Senegal (Bei et al., 2018)</h1>
    <small class="dont-index">Source: <a href="https://github.com/mrc-ide/SIMPLEGEN/blob/HEAD/R/data.R" class="external-link"><code>R/data.R</code></a></small>
    <div class="hidden name"><code>Bei_2018.Rd</code></div>
    </div>

    <div class="ref-description">
    <p>Data from Bei et al. (2018). Here we give a brief summary of the data - see
the original paper for full details.
<br><br>
Samples were randomly selected from longitudinal cohorts collected from
Dielmo/Ndiop, Senegal, and focus on 2 distant time-points (2001-2002, and
2014). The first time-point corresponds to a period of high transmission, and
the second a period of extremely low transmission. Samples were genotyped
using a 24-SNP barcode, and complexity of infection was estimated using the
COIL algorithm. Samples were also matched against a large database of
previously published and unpublished barcodes. For samples from the low
transmission period (2014), one of the three repeated barcode clusters (n =
6) corresponded to a parasite type (haplotype 3), observed in Thiès in both
2007 and 2010. The other 2 clusters (IP1, n = 2; IP2, n = 2) had not
previously been observed. Data were extracted using Tabula v1.2.1.</p>
    </div>

    <div id="ref-usage">
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="va">Bei_2018</span><span class="op">)</span></span></code></pre></div>
    </div>

    <div id="format">
    <h2>Format</h2>
    <p>A list of multiple data objects:</p><ul><li><p><code>EIR</code>: estimates of the EIR at each location and year</p></li>
<li><p><code>barcodes</code>: genetic data and associated sample characteristics</p></li>
<li><p><code>SNP_locations</code>: genomic locations of SNPs (key to be used with <code>barcodes</code>)</p></li>
</ul><p><code>EIR</code>: A dataframe with 3 columns, giving the time, the sampling
  location, and the estimated EIR (see original paper for details). The EIR
  in 2014 in Ndiop was recorded as "0.0" in the paper, and so has been coded
  as "&lt;0.05" here to indicate the precision of this estimate.</p>  
<p><code>barcodes</code>: A dataframe with 34 columns. Gives sample characteristics
  (columns 1:7), the estimated COI and whether this indicates a
  monogenomic/polygenomic infection (columns 8:9), the individual SNPs
  (columns 10:33) and the corresponding haplotype, if known (column 34).</p>  
<p><code>SNP_locations</code>: A dataframe that acts as a key relating the SNP codes
  present in <code>barcodes</code> to the corresponding genomic location.</p>
    </div>
    <div id="source">
    <h2>Source</h2>
    <p><a href="https://academic.oup.com/jid/article/217/4/622/4793403" class="external-link">Data in main paper</a>
  <a href="https://tabula.technology/" class="external-link">Tabula data extractor</a></p>
    </div>
    <div id="references">
    <h2>References</h2>
    <p>Bei AK, Niang M, Deme AB, Daniels RF, Sarr FD, Sokhna C, Talla C, Faye J, Diagne N, Doucoure S, Mboup S, Wirth DF, Tall A, Ndiaye D, Hartl DL, Volkman SK, Toure-Balde A (2018).
“Dramatic Changes in Malaria Population Genetic Complexity in Dielmo and Ndiop, Senegal, Revealed Using Genomic Surveillance.”
<em>The Journal of Infectious Diseases</em>, <b>217</b>(4), 622--627.
ISSN 0022-1899, 1537-6613, doi: <a href="https://doi.org/10.1093/infdis/jix580" class="external-link">10.1093/infdis/jix580</a>
, <a href="https://academic.oup.com/jid/article/217/4/622/4793403" class="external-link">https://academic.oup.com/jid/article/217/4/622/4793403</a>.</p>
    </div>

  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <nav id="toc" data-toggle="toc" class="sticky-top"><h2 data-toc-skip>Contents</h2>
    </nav></div>
</div>


      <footer><div class="copyright">
  <p></p><p>Developed by Bob Verity, Shazia Ruybal, Isobel Routledge, Sophie Berube, Daniel Larremore.</p>
</div>

<div class="pkgdown">
  <p></p><p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.7.</p>
</div>

      </footer></div>

  


  

  </body></html>

