<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1.0"><title>SNP data collected from Uganda (Chang et al., 2017) — Chang_2017 • SIMPLEGEN</title><!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link href="https://cdnjs.cloudflare.com/ajax/libs/bootswatch/3.4.0/flatly/bootstrap.min.css" rel="stylesheet" crossorigin="anonymous"><script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css"><script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous"><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous"><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet"><script src="../pkgdown.js"></script><link href="../extra.css" rel="stylesheet"><meta property="og:title" content="SNP data collected from Uganda (Chang et al., 2017) — Chang_2017"><meta property="og:description" content="Data from Chang et al. (2017). Here we give a brief summary of the data - see
the original paper for full details.


Dried blood spots were taken from 2012-13 cross sectional surveys in several
provinces in Uganda. Households (n = 200) were randomly selected from each
province. All samples that had detectable asexual parasitemia were selected
for Sequenom SNP genotyping. Medium to high frequency SNPs (n = 128) with
high frequency in malaria populations (pf-community-project) were chosen,
leaving 105 SNPs after filtering variants with lower or missing frequencies.
Genotyping was based on the intensity of the SNPs. In addition, merozoite
surface protein 2 (msp2) genotyping was conducted on an age stratified subset
of the samples. Capillary electrophoresis was used to distinguish msp2 allele
sizes.


The data contains a list (size = 3) of data frames. The first data frame
contains SNP calls for 105 filtered SNPs. The values {-1, 0, 0.5, 1} denote
missing value / no call, heterozygous, or homozygous alleles. The second and
third data frame contain 95
frequencies using their described categorical and proportional methods for
modeling homozygous/heterozygous calls and with-in host allele frequency,
respectively."><!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--></head><body data-spy="scroll" data-target="#toc">
    

    <div class="container template-reference-topic">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">SIMPLEGEN</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">0.2.0</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav"><li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Get started
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu"><li>
      <a href="../articles/installation.html">Installation</a>
    </li>
    <li class="divider">
    <li class="dropdown-header">Simulating malaria transmission</li>
    <li>
      <a href="../404.html">A simple model</a>
    </li>
    <li>
      <a href="../404.html">Sampling from the population</a>
    </li>
    <li class="divider">
    <li class="dropdown-header">Overlaying genetics</li>
    <li>
      <a href="../404.html">Pruning the transmission record</a>
    </li>
    <li>
      <a href="../404.html">Pedigrees and relatedness</a>
    </li>
  </ul></li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    The pipeline
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu"><li>
      <a href="../404.html">What is the SIMPLEGEN pipeline?</a>
    </li>
    <li>
      <a href="../articles/transmission_record.html">The transmission record</a>
    </li>
  </ul></li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Models and calibration
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu"><li class="dropdown-header">Overall approach</li>
    <li>
      <a href="../articles/calibration_approach.html">Model development and calibration</a>
    </li>
    <li class="divider">
    <li class="dropdown-header">Transmission models</li>
    <li>
      <a href="../articles/model_description.html">The inbuilt transmission model</a>
    </li>
    <li class="dropdown-submenu">
      <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">Calibration</a>
      <ul class="dropdown-menu" role="menu"><li>
          <a href="../articles/calibration_prev_inc.html">Prevalence, incidence and EIR</a>
        </li>
        <li>
          <a href="../articles/calibration_diagnostics.html">Diagnostics</a>
        </li>
        <li>
          <a href="../articles/calibration_age_inc.html">Age-incidence</a>
        </li>
      </ul></li>
    <li class="divider">
    <li class="dropdown-header">Genetic models</li>
    <li>
      <a href="../404.html">TODO - description</a>
    </li>
    <li class="dropdown-submenu">
      <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">Calibration</a>
      <ul class="dropdown-menu" role="menu"><li>
          <a href="../404.html">TODO - genetic calibration</a>
        </li>
      </ul></li>
  </ul></li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Data
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu"><li class="dropdown-header">Epidemiological data</li>
    <li>
      <a href="../articles/data_prev_inc.html">Prevalence, incidence and EIR</a>
    </li>
    <li>
      <a href="../articles/data_diagnostics.html">Diagnostic sensitivity</a>
    </li>
    <li>
      <a href="../articles/data_age_inc.html">Age-incidence</a>
    </li>
    <li class="divider">
    <li class="dropdown-header">Genetic data</li>
    <li>
      <a href="../404.html">Complexity of infection</a>
    </li>
  </ul></li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Misc
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu"><li>
      <a href="../articles/SLiM_review.html">Integrating with SLiM</a>
    </li>
  </ul></li>
<li>
  <a href="../reference/index.html">Functions</a>
</li>
      </ul><ul class="nav navbar-nav navbar-right"><li>
  <a href="https://github.com/mrc-ide/SIMPLEGEN" class="external-link">
    <span class="fa fa-github"></span>
     
  </a>
</li>
      </ul></div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      

      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>SNP data collected from Uganda (Chang et al., 2017)</h1>
    <small class="dont-index">Source: <a href="https://github.com/mrc-ide/SIMPLEGEN/blob/HEAD/R/data.R" class="external-link"><code>R/data.R</code></a></small>
    <div class="hidden name"><code>Chang_2017.Rd</code></div>
    </div>

    <div class="ref-description">
    <p>Data from Chang et al. (2017). Here we give a brief summary of the data - see
the original paper for full details.
<br><br>
Dried blood spots were taken from 2012-13 cross sectional surveys in several
provinces in Uganda. Households (n = 200) were randomly selected from each
province. All samples that had detectable asexual parasitemia were selected
for Sequenom SNP genotyping. Medium to high frequency SNPs (n = 128) with
high frequency in malaria populations (pf-community-project) were chosen,
leaving 105 SNPs after filtering variants with lower or missing frequencies.
Genotyping was based on the intensity of the SNPs. In addition, merozoite
surface protein 2 (msp2) genotyping was conducted on an age stratified subset
of the samples. Capillary electrophoresis was used to distinguish msp2 allele
sizes.
<br><br>
The data contains a list (size = 3) of data frames. The first data frame
contains SNP calls for 105 filtered SNPs. The values {-1, 0, 0.5, 1} denote
missing value / no call, heterozygous, or homozygous alleles. The second and
third data frame contain 95<!-- % credible intervals for COI and allele -->
frequencies using their described categorical and proportional methods for
modeling homozygous/heterozygous calls and with-in host allele frequency,
respectively.</p>
    </div>

    <div id="ref-usage">
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="va">Chang_2017</span><span class="op">)</span></span></code></pre></div>
    </div>

    <div id="format">
    <h2>Format</h2>
    <p>A list of multiple data objects</p><ul><li><p><code>S1_Table_SNP_data</code>: SNP calls for each sample {-1, 0, 0.5, 1}.</p></li>
<li><p><code>95_cred_interval_coi</code>: Calculated COI of Uganda Samples
    using categorical and proportional methods described in paper.</p></li>
<li><p><code>95_cred_interval_allele_freq</code>: Calculated allele frequency
    of Uganda Samples using categorical and proportional methods described
    in paper.</p></li>
</ul><p><code>S1_Table_SNP_data</code>: A data frame of 107 columns. The sample id and 
 location are included in the first 2 columns. SNP calls (columns 3:107) are 
 either no call / missing data (-1), heterozygous (0.5) or homozygous (0 or 
 1).</p> 
<p><code>S2_Table_95_cred_int_coi_uganda_samples</code>: A data frame of 15 columns 
 containing calculated complexity of infection summary statistics for
 categorical, proportional, and COIL methods.</p> 
<p><code>S3_Table_95_cred_int_allele_freq</code>: A data frame of 25 columns
 containing summary statistics of allele frequency from categorical, 
 proportional, and COIL methods.</p>
    </div>
    <div id="source">
    <h2>Source</h2>
    <p><a href="https://www.ncbi.nlm.nih.gov/pmc/articles/PMC5300274/" class="external-link">Supplementary materials</a></p>
    </div>
    <div id="references">
    <h2>References</h2>
    <p>Chang H, Worby CJ, Yeka A, Nankabirwa J, Kamya MR, Staedke SG, Dorsey G, Murphy M, Neafsey DE, Jeffreys AE, Hubbart C, Rockett KA, Amato R, Kwiatkowski DP, Buckee CO, Greenhouse B (2017).
“THE REAL McCOIL: A method for the concurrent estimation of the complexity of infection and SNP allele frequency for malaria parasites.”
<em>PLoS Comput Biol</em>, <b>13</b>(1), e1005348.
ISSN 1553-734X, doi: <a href="https://doi.org/10.1371/journal.pcbi.1005348" class="external-link">10.1371/journal.pcbi.1005348</a>
, <a href="https://www.ncbi.nlm.nih.gov/pmc/articles/PMC5300274/" class="external-link">https://www.ncbi.nlm.nih.gov/pmc/articles/PMC5300274/</a>.</p>
    </div>

  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <nav id="toc" data-toggle="toc" class="sticky-top"><h2 data-toc-skip>Contents</h2>
    </nav></div>
</div>


      <footer><div class="copyright">
  <p></p><p>Developed by Bob Verity, Shazia Ruybal, Isobel Routledge, Sophie Berube, Daniel Larremore.</p>
</div>

<div class="pkgdown">
  <p></p><p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.7.</p>
</div>

      </footer></div>

  


  

  </body></html>

