<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1.0"><title>24-SNP barcode data showing population structure in Haiti (Charles et al.,
2016) — Charles_2016 • SIMPLEGEN</title><!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link href="https://cdnjs.cloudflare.com/ajax/libs/bootswatch/3.4.0/flatly/bootstrap.min.css" rel="stylesheet" crossorigin="anonymous"><script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css"><script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous"><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous"><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet"><script src="../pkgdown.js"></script><link href="../extra.css" rel="stylesheet"><meta property="og:title" content="24-SNP barcode data showing population structure in Haiti (Charles et al.,
2016) — Charles_2016"><meta property="og:description" content="Data from Charles et al. (2016). Here we give a brief summary of the data -
see the original paper for full details.


Samples were collected through routine surveillance spanning 2006–2009 by the
Haitian Group for the Study of Kaposi’s Sarcoma and Opportunistic Infections
(GHESKIO) at 9 healthcare centers in various municipalities. Samples were
genotyped using a 24-SNP barcode. The paper states that samples that showed
&amp;gt;1 mixed-base SNP call or had &amp;gt;5 missing calls in the 24SNP molecular barcode
were removed from analysis, however, there is 1 sample (monogenomic sample
24) that has 6 missing calls and has not been removed, and there is also 1
sample (polygenomic sample 8) that has exactly 1 mixed-base SNP call and 0
missing calls and yet has been removed. Therefore, it is possible that
samples were removed if they showed &amp;gt;0 mixed-base calls, or had &amp;gt;6 missing
calls (all data presented are consistent with this filtering). Both
monogenomic (n = 52) and polygenomic (n = 8) samples are available, although
samples excluded based on missingness are not available. Samples were
compared in terms of the proportion of shared bases, and were identified as
either identical, related, or unique based on this number."><!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--></head><body data-spy="scroll" data-target="#toc">
    

    <div class="container template-reference-topic">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">SIMPLEGEN</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">0.2.0</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav"><li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Get started
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu"><li>
      <a href="../articles/installation.html">Installation</a>
    </li>
    <li class="divider">
    <li class="dropdown-header">Simulating malaria transmission</li>
    <li>
      <a href="../404.html">A simple model</a>
    </li>
    <li>
      <a href="../404.html">Sampling from the population</a>
    </li>
    <li class="divider">
    <li class="dropdown-header">Overlaying genetics</li>
    <li>
      <a href="../404.html">Pruning the transmission record</a>
    </li>
    <li>
      <a href="../404.html">Pedigrees and relatedness</a>
    </li>
  </ul></li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    The pipeline
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu"><li>
      <a href="../404.html">What is the SIMPLEGEN pipeline</a>
    </li>
    <li>
      <a href="../articles/transmission_record.html">The transmission record</a>
    </li>
  </ul></li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Models and calibration
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu"><li class="dropdown-header">Overall approach</li>
    <li>
      <a href="../articles/calibration_approach.html">Model development and calibration</a>
    </li>
    <li class="divider">
    <li class="dropdown-header">Transmission models</li>
    <li>
      <a href="../articles/model_description.html">The inbuilt transmission model</a>
    </li>
    <li class="dropdown-submenu">
      <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">Calibration</a>
      <ul class="dropdown-menu" role="menu"><li>
          <a href="../articles/calibration_prev_inc.html">Prevalence, incidence and EIR</a>
        </li>
        <li>
          <a href="../articles/calibration_diagnostics.html">Diagnostics</a>
        </li>
        <li>
          <a href="../articles/calibration_age_inc.html">Age-incidence</a>
        </li>
      </ul></li>
    <li class="divider">
    <li class="dropdown-header">Genetic models</li>
    <li>
      <a href="../404.html">TODO - description</a>
    </li>
    <li class="dropdown-submenu">
      <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">Calibration</a>
      <ul class="dropdown-menu" role="menu"><li>
          <a href="../404.html">TODO - genetic calibration</a>
        </li>
      </ul></li>
  </ul></li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Data
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu"><li class="dropdown-header">Epidemiological data</li>
    <li>
      <a href="../articles/data_prev_inc.html">Prevalence, incidence and EIR</a>
    </li>
    <li>
      <a href="../articles/data_diagnostics.html">Diagnostic sensitivity</a>
    </li>
    <li>
      <a href="../articles/data_age_inc.html">Age-incidence</a>
    </li>
    <li class="divider">
    <li class="dropdown-header">Genetic data</li>
    <li>
      <a href="../404.html">Complexity of infection</a>
    </li>
  </ul></li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Misc
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu"><li>
      <a href="../404.html">Integrating with SLiM</a>
    </li>
  </ul></li>
<li>
  <a href="../reference/index.html">Functions</a>
</li>
      </ul><ul class="nav navbar-nav navbar-right"><li>
  <a href="https://github.com/mrc-ide/SIMPLEGEN" class="external-link">
    <span class="fa fa-github"></span>
     
  </a>
</li>
      </ul></div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      

      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>24-SNP barcode data showing population structure in Haiti (Charles et al.,
2016)</h1>
    <small class="dont-index">Source: <a href="https://github.com/mrc-ide/SIMPLEGEN/blob/HEAD/R/data.R" class="external-link"><code>R/data.R</code></a></small>
    <div class="hidden name"><code>Charles_2016.Rd</code></div>
    </div>

    <div class="ref-description">
    <p>Data from Charles et al. (2016). Here we give a brief summary of the data -
see the original paper for full details.
<br><br>
Samples were collected through routine surveillance spanning 2006–2009 by the
Haitian Group for the Study of Kaposi’s Sarcoma and Opportunistic Infections
(GHESKIO) at 9 healthcare centers in various municipalities. Samples were
genotyped using a 24-SNP barcode. The paper states that samples that showed
&gt;1 mixed-base SNP call or had &gt;5 missing calls in the 24SNP molecular barcode
were removed from analysis, however, there is 1 sample (monogenomic sample
24) that has 6 missing calls and has not been removed, and there is also 1
sample (polygenomic sample 8) that has exactly 1 mixed-base SNP call and 0
missing calls and yet has been removed. Therefore, it is possible that
samples were removed if they showed &gt;0 mixed-base calls, or had &gt;6 missing
calls (all data presented are consistent with this filtering). Both
monogenomic (n = 52) and polygenomic (n = 8) samples are available, although
samples excluded based on missingness are not available. Samples were
compared in terms of the proportion of shared bases, and were identified as
either identical, related, or unique based on this number.</p>
    </div>

    <div id="ref-usage">
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="va">Charles_2016</span><span class="op">)</span></span></code></pre></div>
    </div>

    <div id="format">
    <h2>Format</h2>
    <p>A list of two data objects:</p><ul><li><p><code>monoclonal</code>: the 42 samples identified as monoclonal</p></li>
<li><p><code>polyclonal</code>: the remaining 8 samples identified as polyclonal</p></li>
</ul><p><code>monoclonal</code>: A dataframe with 6 columns, giving sample
  characteristics (columns 1:3), barcode data (columns 4:5), and similarity
  category (column 6). location, and the estimated EIR (see original paper
  for details). Heterozygous genotyping calls are identified by "N", and
  missing alleles are identified by "X".</p>  
<p><code>polyclonal</code>: A dataframe with 4 columns, giving sample
  characteristics (columns 1:3), and barcode data (column 4). As above,
  heterozygous genotyping calls are identified by "N", and missing alleles
  are identified by "X".</p>
    </div>
    <div id="source">
    <h2>Source</h2>
    <p><a href="http://wwwnc.cdc.gov/eid/article/22/5/15-0359_article.htm" class="external-link">Data extracted from tables</a></p>
    </div>
    <div id="references">
    <h2>References</h2>
    <p>Charles M, Das S, Daniels R, Kirkman L, Delva GG, Destine R, Escalante A, Villegas L, Daniels NM, Shigyo K, Volkman SK, Pape JW, Golightly LM (2016).
“Plasmodium falciparum K76T pfcrt Gene Mutations and Parasite Population Structure, Haiti, 2006–2009.”
<em>Emerging Infectious Diseases</em>, <b>22</b>(5), 786--793.
ISSN 1080-6040, 1080-6059, doi: <a href="https://doi.org/10.3201/eid2205.150359" class="external-link">10.3201/eid2205.150359</a>
, <a href="http://wwwnc.cdc.gov/eid/article/22/5/15-0359_article.htm" class="external-link">http://wwwnc.cdc.gov/eid/article/22/5/15-0359_article.htm</a>.</p>
    </div>

  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <nav id="toc" data-toggle="toc" class="sticky-top"><h2 data-toc-skip>Contents</h2>
    </nav></div>
</div>


      <footer><div class="copyright">
  <p></p><p>Developed by Bob Verity, Shazia Ruybal, Isobel Routledge, Sophie Berube, Daniel Larremore.</p>
</div>

<div class="pkgdown">
  <p></p><p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.7.</p>
</div>

      </footer></div>

  


  

  </body></html>

