<!-- Generated by pkgdown: do not edit by hand -->
<!DOCTYPE html>
<html lang="en">
  <head>
  <meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">

<title>Draw tree of genome-wide relatedness from pruned transmission record — sim_relatedness • SIMPLEGEN</title>


<!-- jquery -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script>
<!-- Bootstrap -->
<link href="https://cdnjs.cloudflare.com/ajax/libs/bootswatch/3.4.0/flatly/bootstrap.min.css" rel="stylesheet" crossorigin="anonymous" />


<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script>

<!-- bootstrap-toc -->
<link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script>

<!-- Font Awesome icons -->
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous" />
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous" />

<!-- clipboard.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script>

<!-- headroom.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script>

<!-- pkgdown -->
<link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script>




<meta property="og:title" content="Draw tree of genome-wide relatedness from pruned transmission record — sim_relatedness" />
<meta property="og:description" content="Reads in the pruned transmission record from file and creates a
  new node for each inoculation ID. Nodes at time zero are initialised with a
  single haplotype, created de novo with a unique haplotype ID. In subsequent
  generations the nodes that are ancestral to the focal node are known from
  the pruned transmission record. The haplotypes from ancestral nodes are
  sampled at random, and brought together in pairs to produce oocysts. The
  recombinant products of these oocysts are then sampled down to produce a
  new generation of haplotype IDs for this node, along with the relative
  densities of each haplotype." />




<!-- mathjax -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script>

<!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->



  </head>

  <body data-spy="scroll" data-target="#toc">
    <div class="container template-reference-topic">
      <header>
      <div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">SIMPLEGEN</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">0.1.0</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Vignettes
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
    <li>
      <a href="../articles/installation.html">Installation</a>
    </li>
    <li>
      <a href="../articles/basic_tutorial.html">Basic Tutorial</a>
    </li>
    <li class="divider"></li>
    <li>
      <a href="../articles/model_description.html">Transmission Model Details</a>
    </li>
    <li>
      <a href="../articles/transmission_record.html">The Transmission Record</a>
    </li>
    <li>
      <a href="../articles/simulating_relatedness.html">Simulating Relatedness</a>
    </li>
  </ul>
</li>
<li>
  <a href="../reference/index.html">Functions</a>
</li>
      </ul>
      <ul class="nav navbar-nav navbar-right">
        <li>
  <a href="https://github.com/mrc-ide/SIMPLEGEN">
    <span class="fas fa-github"></span>
     
  </a>
</li>
      </ul>
      
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      

      </header>

<div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>Draw tree of genome-wide relatedness from pruned transmission record</h1>
    <small class="dont-index">Source: <a href='https://github.com/mrc-ide/SIMPLEGEN/blob/master/R/main.R'><code>R/main.R</code></a></small>
    <div class="hidden name"><code>sim_relatedness.Rd</code></div>
    </div>

    <div class="ref-description">
    <p>Reads in the pruned transmission record from file and creates a
  new node for each inoculation ID. Nodes at time zero are initialised with a
  single haplotype, created de novo with a unique haplotype ID. In subsequent
  generations the nodes that are ancestral to the focal node are known from
  the pruned transmission record. The haplotypes from ancestral nodes are
  sampled at random, and brought together in pairs to produce oocysts. The
  recombinant products of these oocysts are then sampled down to produce a
  new generation of haplotype IDs for this node, along with the relative
  densities of each haplotype.</p>
    </div>

    <pre class="usage"><span class='fu'>sim_relatedness</span><span class='op'>(</span>
  <span class='va'>project</span>,
  pruned_record_location <span class='op'>=</span> <span class='st'>""</span>,
  r <span class='op'>=</span> <span class='fl'>1e-06</span>,
  alpha <span class='op'>=</span> <span class='fl'>1</span>,
  oocyst_distribution <span class='op'>=</span> <span class='fu'><a href='https://rdrr.io/r/stats/Poisson.html'>dpois</a></span><span class='op'>(</span><span class='fl'>1</span><span class='op'>:</span><span class='fl'>10</span>, lambda <span class='op'>=</span> <span class='fl'>2</span><span class='op'>)</span>,
  hepatocyte_distribution <span class='op'>=</span> <span class='fu'><a href='https://rdrr.io/r/stats/Poisson.html'>dpois</a></span><span class='op'>(</span><span class='fl'>1</span><span class='op'>:</span><span class='fl'>10</span>, lambda <span class='op'>=</span> <span class='fl'>5</span><span class='op'>)</span>,
  contig_lengths <span class='op'>=</span> <span class='fu'><a href='https://rdrr.io/r/base/c.html'>c</a></span><span class='op'>(</span><span class='fl'>643292</span>, <span class='fl'>947102</span>, <span class='fl'>1060087</span>, <span class='fl'>1204112</span>, <span class='fl'>1343552</span>, <span class='fl'>1418244</span>, <span class='fl'>1501717</span>,
    <span class='fl'>1419563</span>, <span class='fl'>1541723</span>, <span class='fl'>1687655</span>, <span class='fl'>2038337</span>, <span class='fl'>2271478</span>, <span class='fl'>2895605</span>, <span class='fl'>3291871</span><span class='op'>)</span>,
  silent <span class='op'>=</span> <span class='cn'>FALSE</span>
<span class='op'>)</span></pre>

    <h2 class="hasAnchor" id="arguments"><a class="anchor" href="#arguments"></a>Arguments</h2>
    <table class="ref-arguments">
    <colgroup><col class="name" /><col class="desc" /></colgroup>
    <tr>
      <th>project</th>
      <td><p>a SIMPLEGEN project, as produced by the
<code><a href='simplegen_project.html'>simplegen_project()</a></code> function.</p></td>
    </tr>
    <tr>
      <th>pruned_record_location</th>
      <td><p>the file path from which the pruned
transmission record will be read.</p></td>
    </tr>
    <tr>
      <th>r</th>
      <td><p>the rate of recombination. The expected number of base pairs in a
single recombinant block is 1/r.</p></td>
    </tr>
    <tr>
      <th>alpha</th>
      <td><p>parameter dictating the skew of haplotype densities. Small
values of <code>alpha</code> create a large skew, and hence make it likely that
an oocyst will be produced from the same parents. Large values of
<code>alpha</code> tend toward more even densities.</p></td>
    </tr>
    <tr>
      <th>oocyst_distribution</th>
      <td><p>vector specifying the probability distribution of
each number of oocysts within the mosquito midgut.</p></td>
    </tr>
    <tr>
      <th>hepatocyte_distribution</th>
      <td><p>vector specifying the probability distribution
of the number of infected hepatocytes in a human host. More broadly, this
defines the number of independent draws from the oocyst products that make
it into the host bloodstream upon a bite from an infectious mosquito.</p></td>
    </tr>
    <tr>
      <th>contig_lengths</th>
      <td><p>vector of lengths (in bp) of each contig.</p></td>
    </tr>
    <tr>
      <th>silent</th>
      <td><p>whether to suppress written messages to the console.</p></td>
    </tr>
    </table>


  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <nav id="toc" data-toggle="toc" class="sticky-top">
      <h2 data-toc-skip>Contents</h2>
    </nav>
  </div>
</div>


      <footer>
      <div class="copyright">
  <p>Developed by Bob Verity, Sophie Berube, Daniel Larremore.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.6.1.</p>
</div>

      </footer>
   </div>

  


  </body>
</html>


